pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to locate network attachment definition %s/%s", namespace, networkName)
+	}
+
+	if nad.Spec.Config == "" {
+		return "", nil
+	}
+
+	netConf := struct {
+		AllowPersistentIPs bool   `json:"allowPersistentIPs,omitempty"`
+		Name               string `json:"name,omitempty"`
+	}{}
+	err = json.Unmarshal([]byte(nad.Spec.Config), &netConf)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal NAD spec.config JSON: %v", err)
+	}
+
+	if !netConf.AllowPersistentIPs {
+		return "", nil
+	}
+
+	if netConf.Name == "" {
+		return "", fmt.Errorf("failed to obtain network name: cannot be empty when allowPersistentIPs is true")
+	}
+
+	return netConf.Name, nil
+}
+
+func CreateIPAMClaims(client kubecli.KubevirtClient, vmi *virtv1.VirtualMachineInstance, ownerRef *v1.OwnerReference) error {
+	nonAbsentNetworks := filterNonAbsentNetworks(vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks)
+	networkToIPAMClaimParams, err := GetNetworkToIPAMClaimParams(client, vmi.Namespace, vmi.Name, nonAbsentNetworks)
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	// For a standalone VMI, the IPAMClaim owner is the VMI itself
+	if ownerRef == nil {
+		ownerRef = v1.NewControllerRef(vmi, virtv1.VirtualMachineInstanceGroupVersionKind)
+	}
+
+	claims := composeIPAMClaims(vmi.Namespace, ownerRef, networkToIPAMClaimParams)
+	err = createIPAMClaims(client, vmi.Namespace, claims)
+	if err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmi.Name, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams.claimName,
+			ipamClaimParams.networkName,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func createIPAMClaims(client kubecli.KubevirtClient, namespace string, claims []*ipamclaims.IPAMClaim) error {
+	for _, claim := range claims {
+		_, err := client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Create(
+			context.Background(),
+			claim,
+			v1.CreateOptions{},
+		)
+
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create IPAMClaim: %w", err)
+			}
+
+			err = validateIPAMClaim(client, namespace, claim.Name, claim.OwnerReferences[0].UID)
+			if err != nil {
+				return fmt.Errorf("failed validating IPAMClaim: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func composeIPAMClaim(namespace string, ownerRef v1.OwnerReference, claimName string, networkName string, interfaceName string) *ipamclaims.IPAMClaim {
```

Comment:
If we pass the vmi.Name we can do the claimName composition here, and is the right place to do it.